morm/iterator.go

```diff
@@ -9,22 +9,16 @@ import (
 )
 
 type ModelIterator interface {
-	// Valid returns whether the current position is valid.
-	// Once invalid, an Iterator is forever invalid.
-	Valid() bool
-
 	// Next moves the iterator to the next sequential key in the database, as
 	// defined by order of iteration.
-	//
-	// If Valid returns false, this method will panic.
-	Next() error
+	Next() (key, value []byte, err error)
 
 	// Load reads the current value at the given key into the passed destination.
 	// This works much like "One" in ModelBucket
-	Load(dest Model) error
+	Load(key, value []byte, dest Model) error
```

Comment:
I think the issue with failing test is somewhere between `Next` and `Load` call. `indexModelIterator` implementation might be the problem. I could not pinpoint the issue though.

Between weave 0.16 and 0.19 we have completely changed the iterator interface and implementation. The previous iterator version was like a pointer - you could inspect the data it points to any time or move one step forward. This was changed into the current version, where the iterator is like a reader. It can only read _once_ the next key-value pair and it is a single, atomic operation.

I think some issues might be caused by having two functions for doing the same task. What do you think about changing `ModelIterator` to 
```go
type ModelIterator interface {
    // Read the next value from the sequence and load the result to given 
    // destination model. Returns errors.ErrIteratorDone if no more data
    // is available.
    LoadNext(destination Model) error
    Release()
}
```
You never want to use `Next` and `Load` separately. Merging them makes the interface smaller and all implementations should be simpler.

If you think this is a good idea, please implement it and let me know if that helped. Otherwise I will think of something else.

Reply:
I was trying to preserve `Next` and `Load` model from `Iterator`, but `ModelIterator` solution seems like much cleaner. Thanks for the tip